src/directLine.ts

```diff
@@ -214,4 +226,92 @@ export class DirectLine implements IBotConnection {
             .delay(5 * 1000)
         );
     }
+
+    /**
+     * Gets an observable of streamURL by requestings a new one from the conversation APIs. 
+     * @return An Observable of stream url.
+     */
+    private getWebSocketStreamURL$(): Observable<string> {
+        return this.connectionStatus$
+            // If connection is OK
+            .filter(connectionStatus => connectionStatus === ConnectionStatus.Online)
+            // Create an Ajax observable
+            .flatMap(_ => {
+                if (this.streamUrl) {
+                    var copy = this.streamUrl;
+                    this.streamUrl = null;
+
+                    return Observable.of(copy);
+                }
+                else {
+                    return Observable.ajax({
+                        method: "GET",
+                        url: `${this.domain}/conversations/${this.conversationId}`,
+                        timeout,
+                        headers: {
+                            "Accept": "application/json",
+                            "Authorization": `Bearer ${this.token}`
+                        }
+                    })// Takes one result
+                    .take(1)
+                    // Extracts the stream URL
+                    .map(result => result.response.streamUrl);
+                }
+            })            
+            // Retries every Timeout seconds in case of connection issue other than security
+            .retryWhen(error$ => error$
+                .mergeMap(error => {
+                    if (error.status === 403) {
+                        this.connectionStatus$.next(ConnectionStatus.Offline);
+                        return Observable.throw(error);
+                    } else {
+                        return Observable.of(error);
+                    }
+                })
+                .delay(timeout)
+            );
+    }
+
+    /**
+     * Gets an Observable of activity from the Web Socket connectivity of direct line.
+     * @return An Observable of activity.
+     */
+    private getWebSocketActivity$(): Observable<Activity> {
+        // From the Web Socket stream URL
+        return this.getWebSocketStreamURL$()
+            // Create an observable Web Socket
+            .flatMap<string, WebSocketResponse>(url => {
+                // Use a result selector to prevent crash from incoming empty message (it could also helps in timeout detection)
+                var ws$ = Observable.webSocket({
+                    url: url,
+                    resultSelector: e => ({                 
+                        type: e.data ? "ACTIVITY" : "PING",
+                        message: e.data ? JSON.parse(e.data) : null
+                    })
+                });
+
+                // timeout seconds ping for keep alive even if already in the browser.
+                Observable.interval(timeout)
+                    .timeInterval()
+                    .subscribe(_ => { ws$.next({}) });
+
```

Comment:
A question about this code: what does it do if you are e.g. under a tunnel? Does WebSocket.send(...) have a notion of timeout? Or will it sit there forever? The reason I ask is,   I don't want this to throw an error on timeout and then go fetch a new URL, because that won't solve the problem.